### PR TITLE
docs: added add method configs

### PR DESCRIPTION
**Pull request by maximilianwerk**
This enriches the Cookbook with information about commonly used `add` parameters. It is still work in progress. Solves #3013

### CODE REVIEW

**Comment by maximilianwerk** [comment]
@alexcg1 @JoanFM @jacobowitz could I get some feedback? or is it ready to be merged from your point of view?

File: .github/2.0/cookbooks/Flow.md [hunk 1]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
 
 </table>
 
+#### Commonly used arguments for distribution setup in `.add`
+
+| Name | default | Description |
+| --- | --- | --- |
+| `host` | `0.0.0.0` | The host of the machine. Can be an ip address or DNS name (e.g. `0.0.0.0`, `my_encoder.jina.ai`) |
+| `port_expose` | randomly initialized | Port of JinaD on the remote machine. |
```
Comment (JoanFM) [inline .github/2.0/cookbooks/Flow.md:439]: Are we sure about this? I think this is only for `gateway` and is the port_exposed by it.

File: .github/2.0/cookbooks/Flow.md [hunk 2]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+| `port_in` | randomly initialized | Port for incoming traffic for the Executor. |
+| `port_out` | randomly initialized | Port for outgoing traffic for the Executor. This is only used in the remote-local use-case described below. |
+| `connect_to_predecessor` | `False` | Forces a Head to connect to the previous Tail. This is only used in the remote-local use-case described below. |
+| `use_external` | `False` | Stops `Flow` from context managing an Executor. This allows spawning of an external Executor and reusing across multiple Flows. |
```
Comment (JoanFM) [inline .github/2.0/cookbooks/Flow.md:443]: Isn't `external` the name of the argument?

File: .github/2.0/cookbooks/Flow.md [hunk 3]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+
+##### Using an already spawned Executor:
+
+The external Executor in the following two use-cases could have been spawned
+
+- either by another Flow
+- or by the `jina executor` CLI command
+
+```python
+f.add(host='localhost', port_in=12345, use_external=True)
```
Comment (JoanFM) [inline .github/2.0/cookbooks/Flow.md:483]: same here

**Review by JoanFM — DISMISSED** [review]

File: .github/2.0/cookbooks/Flow.md (hunk 1 above)
Comment (maximilianwerk) [reply .github/2.0/cookbooks/Flow.md:439]: @deepankarm

File: .github/2.0/cookbooks/Flow.md [hunk 4]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+
+| Name | default | Description |
+| --- | --- | --- |
+| `host` | `0.0.0.0` | The host of the machine. Can be an ip address or DNS name (e.g. `0.0.0.0`, `my_encoder.jina.ai`) |
+| `port_expose` | randomly initialized | Port of JinaD on the remote machine. |
+| `port_in` | randomly initialized | Port for incoming traffic for the Executor. |
+| `port_out` | randomly initialized | Port for outgoing traffic for the Executor. This is only used in the remote-local use-case described below. |
+| `connect_to_predecessor` | `False` | Forces a Head to connect to the previous Tail. This is only used in the remote-local use-case described below. |
+| `external` | `False` | Stops `Flow` from context managing an Executor. This allows spawning of an external Executor and reusing across multiple Flows. |
+| `uses`, `uses_before` and `uses_after` prefix | No prefix | When prefixing one of the `uses` arguments with `docker` or `jinahub+docker`, the Pod does not run natively, but is spawned inside a container. |
```
Comment (alexcg1) [inline .github/2.0/cookbooks/Flow.md:444]: Should this be Pod or Executor?

File: .github/2.0/cookbooks/Flow.md [hunk 5]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+
+#### Commonly used patterns for `.add`
+
+##### Local native Executor via `.py` file:
+
+```python
+class MyExecutor():...
+
+f.add(uses=MyExecutor)
+f.add(uses='MyExecutor')
```
Comment (alexcg1) [inline .github/2.0/cookbooks/Flow.md:454]: This is just adding from a string? I see the same line duplicated except one has quotes

File: .github/2.0/cookbooks/Flow.md [hunk 6]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+```
+
+##### Local native Executor via `.yml` file:
+
+```python
+f.add(uses='mwu_encoder.yml')
+```
+
+##### Local docker Executor:
```
Comment (alexcg1) [inline .github/2.0/cookbooks/Flow.md:463]: **D**ocker

File: .github/2.0/cookbooks/Flow.md [hunk 7]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+
+```python
+f.add(uses='docker://MyExecutor')
+```
+
+##### Local docker Executor from jinahub:
```
Comment (alexcg1) [inline .github/2.0/cookbooks/Flow.md:469]: **D**ocker

File: .github/2.0/cookbooks/Flow.md [hunk 8]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+##### Remote Executor
+
+In any of the following remote patterns without `external`, JinaD must be running on the remote machine.
+Furthermore, JinaD must be setup to listen on `port_expose`.
+
+```python
+f.add(uses='mwu_encoder.yml', host='123.45.67.89', port_in=12345, port_expose=8080)
+```
+
+`uses` can take any argument as in the local case (e.g. `jinahub+docker:://...`).
```
Comment (alexcg1) [inline .github/2.0/cookbooks/Flow.md:496]: `uses` can take any argument that would be used in the local case (e.g. `jinahub+docker:://...`).

File: .github/2.0/cookbooks/Flow.md [hunk 9]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+
+##### Setting the Gateway data port and host
+
+Whenever the networking of the Gateway should be changed, the respective arguments go directly into the Flow initialization.
```
Comment (alexcg1) [inline .github/2.0/cookbooks/Flow.md:500]: ~Whenever the networking of the Gateway should be changed~ If you need to change the networking of the Gateway

File: .github/2.0/cookbooks/Flow.md [hunk 10]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+
+```python
+Flow(host='123.45.67.89', port_in=12345)
+```
+
+##### Forcing an Executor in the remote-local configuration
+
+Sometimes you want to use a remote Executor in your local Flow (e.g. using an expensive encoder on a remote GPU).
+Then the remote cannot talk back to the next local Executor directly.
+This is similar to a server, that cannot talk to a client, before the client has opened a connection.
```
Comment (alexcg1) [inline .github/2.0/cookbooks/Flow.md:510]: Remove the two commas here

File: .github/2.0/cookbooks/Flow.md (hunk 7 above)
Comment (alexcg1) [reply .github/2.0/cookbooks/Flow.md:469]: Jina Hub

File: .github/2.0/cookbooks/Flow.md [hunk 11]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
+The Flow inside Jina has an auto-detection mechanism for such cases.
+Anyhow, in some networking setups this mechanism fails.
+Then you can force this by hand by setting the `connect_to_predecessor` argument and `port_out` to the Executor in front.
+
+```python
+Flow().add(name='remote', host='123.45.67.89', port_out=23456).add(name='local', connect_to_predecessor=True)
+```
+
+
 ### Send Data Request via `post`
```
Comment (alexcg1) [inline .github/2.0/cookbooks/Flow.md:520]: use sentence case here

**Review by alexcg1 — CHANGES_REQUESTED** [review]
Looking good! A few small changes to make

File: .github/2.0/cookbooks/Flow.md (hunk 5 above)
Comment (maximilianwerk) [reply .github/2.0/cookbooks/Flow.md:454]: Yes you can either pass the object or the name of the Executor.

File: .github/2.0/cookbooks/Flow.md [hunk 12]
```diff
@@ -431,6 +431,92 @@ f = (Flow()
...
 
+#### Commonly used arguments for distribution setup in `.add`
+
+| Name | default | Description |
+| --- | --- | --- |
+| `host` | `0.0.0.0` | The host of the machine. Can be an ip address or DNS name (e.g. `0.0.0.0`, `my_encoder.jina.ai`) |
+| `port_expose` | randomly initialized | Port of JinaD on the remote machine. |
+| `port_in` | randomly initialized | Port for incoming traffic for the Executor. |
+| `port_out` | randomly initialized | Port for outgoing traffic for the Executor. This is only used in the remote-local use-case described below. |
+| `connect_to_predecessor` | `False` | Forces a Head to connect to the previous Tail. This is only used in the remote-local use-case described below. |
```
[the remark this answers is not in the document]
Comment (maximilianwerk) [reply .github/2.0/cookbooks/Flow.md:442]: Well, if you have some private VPN in the cloud, that we cannot detect, the user needs to use this parameter. The question is, whether the usage is described detailed enough in the section below.

File: .github/2.0/cookbooks/Flow.md (hunk 11 above)
Comment (maximilianwerk) [reply .github/2.0/cookbooks/Flow.md:520]: Haha was there already beforehand, but I happily change it.

File: .github/2.0/cookbooks/Flow.md (hunk 1 above)
Comment (maximilianwerk) [reply .github/2.0/cookbooks/Flow.md:439]: But it does not default to 8000. It is randomly initialized. :D I think we need to clarify this in code and ideally split `port_expose` into two concepts:

1. The port the client can speak to
2. The port that JinaD exposes

Comment (maximilianwerk) [reply .github/2.0/cookbooks/Flow.md:454]: I suggest we do the renaming of the parameters in another PR.

File: .github/2.0/cookbooks/Flow.md (hunk 5 above)
Comment (alexcg1) [reply .github/2.0/cookbooks/Flow.md]: Let's make that clear in the doc (unless I missed it somewhere)

Comment (maximilianwerk) [reply .github/2.0/cookbooks/Flow.md]: I did in the latest change. :+1: Thanks for pointing me at it.

**Review by hanxiao — APPROVED** [review]

**Pull request merged**